include/nbl/builtin/hlsl/workgroup/bitonic_sort.hlsl

```diff
@@ -0,0 +1,253 @@
+    {
+        keyAdaptor.template set<Key>(ownedIdx, p.first);
+        valueAdaptor.template set<Value>(ownedIdx, p.second);
+
+        // Wait until all writes are done before reading - only barrier on one adaptor here
+        keyAdaptor.workgroupExecutionAndMemoryBarrier();
+
+        keyAdaptor.template get<Key>(ownedIdx ^ mask, p.first);
+        valueAdaptor.template get<Value>(ownedIdx ^ mask, p.second);
+    }
+    
+
+    template<typename SharedMemoryAccessor>
+    static void mergeStage(NBL_REF_ARG(SharedMemoryAccessor) sharedmemAccessor, uint32_t stage, bool bitonicAscending, uint32_t invocationID,
+        NBL_REF_ARG(nbl::hlsl::pair<key_t, value_t>) lopair, NBL_REF_ARG(nbl::hlsl::pair<key_t, value_t>) hipair)
+    {
+        const uint32_t WorkgroupSize = config_t::WorkgroupSize;
+        const uint32_t subgroupSizeLog2 = glsl::gl_SubgroupSizeLog2();
+        comparator_t comp;
+
+
+        using key_adaptor_t = accessor_adaptors::StructureOfArrays<SharedMemoryAccessor, uint32_t, uint32_t, 1, WorkgroupSize>;
+        using value_adaptor_t = accessor_adaptors::StructureOfArrays<SharedMemoryAccessor, uint32_t, uint32_t, 1, WorkgroupSize, integral_constant<uint32_t, WorkgroupSize * sizeof(key_t) / sizeof(uint32_t)> >;
+
+        key_adaptor_t keyAdaptor;
+        keyAdaptor.accessor = sharedmemAccessor;
+        value_adaptor_t valueAdaptor;
+        valueAdaptor.accessor = sharedmemAccessor;
+
+        [unroll]
+        for (uint32_t pass = 0; pass <= stage; pass++)
+        {
+            if (pass)
+                sharedmemAccessor.workgroupExecutionAndMemoryBarrier();
+
+            const uint32_t stridePower = (stage - pass + 1) + subgroupSizeLog2;
+            const uint32_t stride = 1u << stridePower;
+            const uint32_t threadStride = stride >> 1;
+
+            nbl::hlsl::pair<key_t, value_t> plopair = lopair;
+            shuffleXor(plopair, invocationID, threadStride, keyAdaptor, valueAdaptor);
+
+            nbl::hlsl::pair<key_t, value_t> phipair = hipair;
+            shuffleXor(phipair, invocationID ^ threadStride, threadStride, keyAdaptor, valueAdaptor);
+
+            const bool isUpper = (invocationID & threadStride) != 0;
+            const bool takeLarger = isUpper == bitonicAscending;
+
+            nbl::hlsl::bitonic_sort::compareExchangeWithPartner(takeLarger, lopair, plopair, hipair, phipair, comp);
+        }
+    }
+
+    template<typename Accessor, typename SharedMemoryAccessor NBL_FUNC_REQUIRES(bitonic_sort::BitonicSortAccessor<Accessor, key_t, value_t>&& bitonic_sort::BitonicSortSharedMemoryAccessor<SharedMemoryAccessor>)
+    static void __call(NBL_REF_ARG(Accessor) accessor, NBL_REF_ARG(SharedMemoryAccessor) sharedmemAccessor)
+    {
+        const uint32_t WorkgroupSize = config_t::WorkgroupSize;
+
+        const uint32_t invocationID = glsl::gl_LocalInvocationID().x;
```

Comment:
oh btw the Spec doesn't say that `LocalInvocationIndex = SubgroupID*SubgroupSize+SubgroupInvocationID`

This is why in such shaders you need to always use https://github.com/Devsh-Graphics-Programming/Nabla/blob/b7f3dfb7ed197b90e5e1877b97a5cfe5e7b62d6f/include/nbl/builtin/hlsl/workgroup/basic.hlsl#L26

instead of `gl_LocalInvocationID` or `gl_LocalInvocationIndex`